python/lsst/ts/tcpip/base_one_client_read_loop_server_test_case.py

```diff
@@ -0,0 +1,76 @@
+# This file is part of ts_tcpip.
+#
+# Developed for the Rubin Observatory Telescope and Site System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import abc
+import asyncio
+import contextlib
+import logging
+import typing
+import unittest
+
+from .client import Client
+from .one_client_read_loop_server import OneClientReadLoopServer
+
+__all__ = ["BaseOneClientReadLoopServerTestCase"]
+
+# Standard timeout in seconds.
+TIMEOUT = 60
+
+# After how many loops should the connection fail?
+FAIL_INDEX = 2
+
+
+class BaseOneClientReadLoopServerTestCase(unittest.IsolatedAsyncioTestCase):
+    async def asyncSetUp(self) -> None:
+        self.log = logging.getLogger(type(self).__name__)
+        self.srv = await self.create_server()
+
+        await self.srv.start_task
+        assert self.srv.server.is_serving()
```

Comment:
I worry about using asyncSetUp to do anything fancy because it is prone to hanging -- thus I always try to use async context managers.

I think it suffices to do this:
```
def setUp():
    self.log = logging.getLogger()

@contextlib.asynccontextmanager
@abc.abstractmethod
async def create_server(self) -> OneClientServer:
    raise NotImplementedError()

async def create_client(self, server) -> Client:
    async with Client(host=server.host, port=server.port, log=self.log) as client:
        await asyncio.wait_for(server.connected_task, timeout=TIMEOUT)
        assert client.connected
        yield client
```
Then test code would do things like this:
```
@contextmanager.asynccontextmanager
def create_server(self) -> OneClientServer:
    async with MyServerSubclass(host=LOCALHOST, port=0, log=self.log) as server:
        yield server

def test_something(self) -> None:
    async with self.create_server() as server, self.create_client(server) as client:
        # do stuff with server and client
```

There isn't all that much left in this base test class, but it might be enough to justify existence. It certainly avoids test cases from having to deal with the host, port, and log arguments.

You can get fancier by making make_server concrete and having it call an abstract method (that's what BaseCscTestCase does with make_csc -- it calls abstract method basic_make_csc). This is a help in BaseCscTestCase because the concrete method also does a lot of extra work. Here, I'm not sure it would be much help -- unless you also want to construct the client (which would probably be OK if all test cases need a client -- in which case you can ditch the create_client method).